src/driver/xdma_sgdma.c

```diff
@@ -261,22 +149,59 @@ ssize_t xdma_channel_read_write(struct xdma_channel *chnl,
 		return -EINVAL;
 	}
 
-	rv = check_transfer_align(engine, buf, count, *pos, 1);
-	if (rv) {
-		pr_info("Invalid transfer alignment detected\n");
-		return rv;
-	}
+	if (sg_alloc_table(&sgt, SG_MAX_SINGLE_ALLOC, GFP_KERNEL))
+		return -ENOMEM;
 
-	memset(&cb, 0, sizeof(struct xdma_io_cb));
-	cb.buf = (char __user *)buf;
-	cb.len = count;
-	rv = xdma_channel_map_user_buf_to_sgl(&cb, write);
-	if (rv < 0)
-		return rv;
+	while (iov_iter_count(iter)) {
+		pages = (struct page **) sgt.sgl;
+		pages += SG_MAX_SINGLE_ALLOC * (PAGE_PTRS_PER_SGL - 1);
+
+		size = iov_iter_get_pages(iter, pages, LONG_MAX,
+				SG_MAX_SINGLE_ALLOC, &offset);
+		if (size < 0) {
+			res = size;
+			goto out;
+		}
+
+		rv = check_transfer_align(engine, (void *) offset, size, pos, 1);
+
```

Comment:
Nit: extra space. Also rv is an odd name. rc or ret are usually conventionally used.